### ISSUE TITLE
Dataflow runner does not respect ParDo's lifecycle on case of exceptions

### ISSUE THREAD

**Issue by kennknowles** [apache/beam#18592]
The lifecycle of the DoFn is not respected in case of exception in any of the lifecycle methods after setup.

Imported from Jira [BEAM-3245](https://issues.apache.org/jira/browse/BEAM-3245). Original Jira may contain additional context.
Reported by: iemejia.

**Comment by Abacn** [comment]
Found that tearDown is never called in Dataflow runner v1 or v2, even pipeline finished normally

Just use a simple DoFn:

```
    pipeline
        .apply(GenerateSequence.from(0).to(1000))
        .apply(ParDo.of(new SomeDoFn()));
```

where SomeDoFn is 

```
static class SomeDoFn extends DoFn<Long, Long> {

    protected transient Integer id;

    @Setup
    public void setup() {
      Integer idOld = id;
      id = ThreadLocalRandom.current().nextInt(0, 1000);
      LOG.info("@Setup {}, {} at {}", idOld, id, Integer.toHexString(System.identityHashCode(this)));
    }

    @StartBundle
    public void startBundle() {
      LOG.info("@StartBundle {} at {}", id, Integer.toHexString(System.identityHashCode(this)));
    }

    @ProcessElement
    public void process(@Element Long input, OutputReceiver<Long> receiver) {
      // LOG.info("@ProcessElement for {} at bundle {}", input, id);
      receiver.output(input);
    }

    @FinishBundle
    public void finishBundle() throws Exception {
      LOG.info("@FinishBundle {} at {}", id, Integer.toHexString(System.identityHashCode(this)));
    }

    @Teardown
    public void teardown() {
      LOG.info("@Teardown {} at {}", id, Integer.toHexString(System.identityHashCode(this)));
    }

    @Override
    protected void finalize() throws Throwable {
      LOG.info("finalize");
    }
  }
```

seen log in setup, start/finishBundle, but not teardown.

**Comment by Abacn** [comment]
Checked that for Dataflow legacy runner, teardown is called after a DoFn throw error in processElement or finishBundle, but not Dataflow runner v2.

**Comment by scwhittle** [comment]
One previously identified issue was that teardown may not be invoked on a DoFn if there is an exception calling setup on a different fused DoFn.

However I believe that if the dataflow runner abandons a MapTaskExecutor for some reason not related to an exception in user processing (cache timeout or exception during backend commit) that we may not be invoking the teardown method.  This can lead to leaking resources if they are created in setup and require teardown to close them.  An example of this is the KafkaProducer in KafkaWriter  [here](sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java#L57).

**Comment by Abacn** [comment]
> One previously identified issue was that teardown may not be invoked on a DoFn if there is an exception calling setup on a different fused DoFn.

Yes, link here: https://github.com/apache/beam/pull/32522/files#r1777301355

> However I believe that if the dataflow runner abandons a MapTaskExecutor for some reason not related to an exception in user processing (cache timeout or exception during backend commit) that we may not be invoking the teardown method. This can lead to leaking resources if they are created in setup and require teardown to close them. An example of this is the KafkaProducer in KafkaWriter [here](sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java#L57).

The link not working, pasted a perma-link here for reference: 

https://github.com/apache/beam/blob/c22665c5111709a9181cf5a620339326ba97e88b/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java#L57

**Comment by scwhittle** [comment]
I put together a PR to fix and will be working next on a test that verifies this is the case in Dataflow Streaming runner but it certainly appears to be so from looking at the code.

**Comment by scwhittle** [comment]
@Abacn Are there remaining issues here that are known or can I close this?
